dbclient/MLFlowClient.py

```diff
@@ -198,3 +192,176 @@ def _cleanse_artifact_location(self, artifact_location):
             return None
         return artifact_location
+
+    def _cleanse_tags(self, tags_map):
+        return dict(filter(lambda key_value: not key_value[0].startswith("mlflow."), tags_map.items()))
+
+    def import_mlflow_runs(self, src_client_config, log_sql_file='mlflow_runs.db', experiment_id_map_log='mlflow_experiments_id_map.log', run_id_map_log='mlflow_runs_id_map.log', ml_run_artifacts_dir='ml_run_artifacts/', num_parallel=4):
+        """
+        Imports the Mlflow run objects. This can be run only after import_mlflow_experiments is complete.
+        Input files are mlflow_runs.db, mlflow_experiments_id_map.log
+        Outputs mlflow_runs_id_map.log which has the map of old_run_id -> new_run_id after imports.
+        """
+        src_client = MlflowClient(f"databricks://{src_client_config['profile']}")
+        experiment_id_map = self._load_experiment_id_map(self.export_dir + experiment_id_map_log)
+        mlflow_runs_file = self.export_dir + log_sql_file
+        os.makedirs(self.export_dir + ml_run_artifacts_dir, exist_ok=True)
+
+        error_logger = logging_utils.get_error_logger(
+            wmconstants.WM_IMPORT, wmconstants.MLFLOW_RUN_OBJECT, self.export_dir
+        )
+
+        # checkpoint is required since the checkpoint file is copied into mlflow_runs_id_map.log at the end of the step.
+        assert self._checkpoint_service.checkpoint_enabled, "import_mlflow_runs requires --use-checkpoint to be enabled. If " \
+                                                            " you need to actually rerun, remove the corresponding " \
+                                                            "checkpoint file. e.g. logs/checkpoint/import_mlflow_runs.log"
+
+        mlflow_runs_checkpointer = self._checkpoint_service.get_checkpoint_key_map(
+            wmconstants.WM_IMPORT, wmconstants.MLFLOW_RUN_OBJECT)
+
+        # This checkpointer is used to checkpoint individual steps for more optimal checkpointing.
+        # e.g. checkpoint run_creation, log_batch, and artifact download_upload separately
+        mlflow_runs_steps_checkpointer = self._checkpoint_service.get_checkpoint_key_map(
+            wmconstants.WM_IMPORT, wmconstants.MLFLOW_RUN_OBJECT + "_steps"
+        )
+
+        start = timer()
+
+        con = sqlite3.connect(mlflow_runs_file)
+        cur = con.execute("SELECT * FROM runs")
+        # TODO(kevin): make this configurable later
+        runs = cur.fetchmany(10000)
+        while(len(runs) > 0):
+            with ThreadPoolExecutor(max_workers=num_parallel) as executor:
+                # run_id = run[0]
+                # start_time = run[1]
+                # run_obj = json.loads(run[2])
+                futures = [executor.submit(self._create_run_and_log, src_client, mlflow_runs_file, run[0], run[1], json.loads(run[2]), experiment_id_map, self.export_dir + ml_run_artifacts_dir, error_logger, mlflow_runs_checkpointer, mlflow_runs_steps_checkpointer) for run in runs]
```

Comment:
BTW out of curiosity why is the run-based locking necessary? I would expect that when using `ThreadPoolExecutor` no two threads would ever attempt to migrate the same run

Reply:
It seems we could handle that by grouping together trees of runs by parent ID first before attempting to import them (so that `r1` is guaranteed to be imported before `r2`). We could also then have an individual thread handle importing a tree of runs without worrying about race conditions potentially. But the current approach seems ok too since it does guarantee that parents are always created before children

Reply:
Yes your approach would work too, but the problem is that `grouping together trees of runs by parent ID first` is not that easy in terms of the performance. In order to do that, we would have to go through all the runs first and **load it in to memory**. Note that we are already chunking the fetchmany by only certain number, but there is no guarantee that all the same rooted trees are in the same chunk. This requires us to load all the data in memory to do the grouping.

Another similar approach is to recognize this at the export time and indicate it in database schema. (by setting a group or something). We can set the list of group Ids in a separate file, and on import, we can probably parallelize across those different group Ids. (where each of the thread does SELECT * FROM RUNS WHERE GROUP_ID=$GROUP_ID).

The current method is complicated but good in a sense that program can do blind parallelization and just wait if it ever runs into a situation where its parernt run is being handled by other threads. The key is to ensure the atomicity of `check_contains_otherwise_mark_in_use` function which is true by testing and the document explanation.